### PR TITLE
[java] FinalFieldCouldBeStatic - fix false negative with unnecessary parenthesis

### LINKED ISSUE #3949: A false negative about the rule FinalFieldCouldBeStatic
**Affects PMD Version:6.42.0** 

**Rule:FinalFieldCouldBeStatic**

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/latest/pmd_rules_java_design.html#finalfieldcouldbestatic>

**Description:**
Hi, I found a false negative about the rule FinalFieldCouldBeStatic. Please read the example below. PMD should have reported a warning in this line, but it doesn't. I think this is caused by the ignorance of parenthesized literal. Thanks for your consideration.

**Code Sample demonstrating the issue:**
```java
public final int BAR = (42);
```

**Expected outcome:**

PMD should report a violation at line 1, but doesn't. This is a false-negative.

**Running PMD through:** *[Maven]*

### PR DESCRIPTION
**Pull request by dalizi007**
## Describe the PR

Fix false negative about the rule FinalFieldCouldBeStatic caused by the ignorance of parenthesized literal
## Related issues

- Fixes #3949

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 380 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/72ad5122565847f49a23f3b03498a31d/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/design.xml [hunk 1]
```diff
@@ -813,14 +813,19 @@ in each object at runtime.
  [@Final=true() and @Static=false()]
  [not(preceding-sibling::Annotation/MarkerAnnotation/Name[@Image="Builder.Default"]
     and //ImportDeclaration/Name[@Image="lombok.Builder"])]
+ [not(ReferenceType)]
 /VariableDeclarator
  [VariableInitializer/Expression/PrimaryExpression[not(PrimarySuffix)]
-  /PrimaryPrefix/*
+  /PrimaryPrefix[not(./AllocationExpression/ClassOrInterfaceType[pmd-java:typeIs('java.lang.Thread')])]
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/design.xml:819]: Is this change related (excluding java.lang.Thread)? If so, please add a test case

File: pmd-java/src/main/resources/category/java/design.xml [hunk 2]
```diff
@@ -813,14 +813,19 @@ in each object at runtime.
...
+    /*
     [
         self::Literal (: literal :)
         or
         (: another static field :)
         self::Name[@Image=//FieldDeclaration[@Static=true()]/VariableDeclarator/@Name]
+        or
+        (:unnecessary parenthesis :)
+        descendant-or-self::Literal
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/design.xml:828]: This `descendant-or-self::Literal` includes `self::Literal`, so the first check in the original implementation can be removed

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for the PR!

Looking at the regression tester reports (https://chunk.io/pmd/72ad5122565847f49a23f3b03498a31d/diff1/index.html) I see a couple of false positives:
* https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4841indentation/InputIndentationCorrectNewChildren.java#L5
    * I don't think the rule should flag this - sure: the whole expression contains lots of literals. In this special case, you probably don't want to make the StringBuffer static (while the reference is a constant, the StringBuffer itself is mutable).
    * We should keep that rule simple and try to only flag direct literals (like in the example)
* https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/FileImportControlTest.java#L29
    * Similar case - I think, we should not consider constructor calls or method calls or anything the executes at runtime.
* https://github.com/checkstyle/checkstyle/tree/checkstyle-9.1/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumber_1.java#L151
    * This is actually a correct - this could be made static. I'm not sure however, whether we should consider these boxing types explictly (would make the rule more complex). It would be fixing a different bug, so maybe it's better to not consider these special cases.
* https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/com/sun/crypto/provider/BlowfishCrypt.java#L283
    * This is also a false positive - it's creating a new array at runtime and is therefore not a compile-time constant
* https://github.com/spring-projects/spring-framework/tree/v5.3.13/spring-aop/src/main/java/org/springframework/aop/aspectj/RuntimeTestWalker.java#L268
    * That could be actually correct - an array with size 0 can't be modified anymore and could be considered a compile time constant. However, like for `new Integer(27)`, this would be fixing a different bug, so maybe it's better to not consider these special cases.

**Pull request merged**